cmd/postgres_exporter/postgres_exporter.go

```diff
@@ -1425,8 +1425,85 @@ func main() {
 		exporter.servers.Close()
 	}()
 
-	prometheus.MustRegister(exporter)
+	exporter_shared.RunServer("PostgreSQL", *listenAddress, *metricPath, newHandler(map[string]prometheus.Collector{
+		"exporter": exporter,
+	}))
+}
+
+// handler wraps an unfiltered http.Handler but uses a filtered handler,
+// created on the fly, if filtering is requested. Create instances with
+// newHandler. It used for collectors filtering.
+type handler struct {
+	unfilteredHandler http.Handler
+	collectors        map[string]prometheus.Collector
+}
+
+func newHandler(collectors map[string]prometheus.Collector) *handler {
+	h := &handler{collectors: collectors}
+	if innerHandler, err := h.innerHandler(); err != nil {
+		log.Fatalf("Couldn't create metrics handler: %s", err)
+	} else {
+		h.unfilteredHandler = innerHandler
+	}
+	return h
+}
+
+// ServeHTTP implements http.Handler.
+func (h *handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	filters := r.URL.Query()["collect[]"]
+	log.Debugln("collect query:", filters)
+
+	if len(filters) == 0 {
+		// No filters, use the prepared unfiltered handler.
+		h.unfilteredHandler.ServeHTTP(w, r)
+		return
+	}
+
+	filteredHandler, err := h.innerHandler(filters...)
+	if err != nil {
+		log.Warnln("Couldn't create filtered metrics handler:", err)
+		w.WriteHeader(http.StatusBadRequest)
+		w.Write([]byte(fmt.Sprintf("Couldn't create filtered metrics handler: %s", err)))
+		return
+	}
+	filteredHandler.ServeHTTP(w, r)
+}
+
+func (h *handler) innerHandler(filters ...string) (http.Handler, error) {
+	// clear all previously registered collector (to prevent repeatable registration).
+	for k, v := range h.collectors {
+		if ok := prometheus.Unregister(v); ok {
```

Comment:
There will be a race if this singleton handler will be scraped concurrently.

We can use disposably registry/gatherer instead of `prometheus.DefaultGatherer`.

Reply:
I do not exactly understand where it will be? In the `Unregister` method? 
It seems it won't, as documentation said: [All exported functions and methods are safe to be used concurrently unless specified otherwise.](https://godoc.org/github.com/prometheus/client_golang/prometheus). But ok. I'll replace this by separate `registry`.

Reply:
> I do not exactly understand where it will be? In the Unregister method?

Unregister and Register methods are thread-safe. Their usage, however, is not synchronized in this code. Two concurrent scrapes will concurrently mutate the shared registry resulting in different results.

> Or do you mean in the for cycle, where I range over the map?

map is not being mutated.

Reply:
> Their usage, however, is not synchronized in this code. Two concurrent scrapes will concurrently mutate the shared registry resulting in different results.

Make sense.